src/Features/LanguageServer/Protocol/Features/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer_GetDiagnosticsForSpan.cs

```diff
@@ -357,24 +371,43 @@ public async Task<bool> TryGetAsync(ArrayBuilder<DiagnosticData> list, Cancellat
                             list.Add(item);
                     }
 
-                    return true;
+                    return (true, existingData);
                 }
 
-                return false;
+                return (false, existingData);
             }
 
             private async Task ComputeDocumentDiagnosticsAsync(
-                ImmutableArray<StateSet> stateSets,
+                ImmutableArray<(StateSet stateSet, DocumentAnalysisData? existingData)> stateSetsAndExistingData,
                 AnalysisKind kind,
                 TextSpan? span,
                 ArrayBuilder<DiagnosticData> builder,
                 bool incrementalAnalysis,
                 CancellationToken cancellationToken)
             {
                 Debug.Assert(!incrementalAnalysis || kind == AnalysisKind.Semantic);
-                Debug.Assert(!incrementalAnalysis || stateSets.All(stateSet => stateSet.Analyzer.SupportsSpanBasedSemanticDiagnosticAnalysis()));
+                Debug.Assert(!incrementalAnalysis || stateSetsAndExistingData.All(stateSetAndData => stateSetAndData.stateSet.Analyzer.SupportsSpanBasedSemanticDiagnosticAnalysis()));
+
+                using var _ = ArrayBuilder<StateSet>.GetInstance(stateSetsAndExistingData.Length, out var stateSetBuilder);
+                foreach (var stateSetAndData in stateSetsAndExistingData)
+                {
+                    var analyzer = stateSetAndData.stateSet.Analyzer;
+                    if (_priorityProvider.MatchesPriority(analyzer))
+                    {
+                        // Check if this is an expensive analyzer that needs to be de-prioritized to a lower priority bucket.
+                        // If so, we skip this analyzer from execution in the current priority bucket.
+                        // We will subsequently execute this analyzer in the lower priority bucket.
+                        Contract.ThrowIfNull(stateSetAndData.existingData);
+                        if (await TryDeprioritizeAnalyzerAsync(analyzer, stateSetAndData.existingData.Value).ConfigureAwait(false))
```

Comment:
This is the core logic addition to deprioritize analyzer to 'Low' bucket based on the conditions we discussed offline.